### ISSUE TITLE
Rename Slip10KeyringEntry -> Slip10Wallet

### ISSUE THREAD

**Issue by webmaster128** [iov-one/iov-core#384]
as discussed in #302. Depends on #379 and #380 to avoid unnecessary merge conflicts.

This is non-breaking since Slip10KeyringEntry is not part of the public API.

**Comment by SpasZahariev** [comment]
I want to try, is there anything else I need to consider?

**Comment by ethanfrey** [comment]
Hi @SpasZahariev 

first off, welcome to iov-core and thank you for the offer.

I would actually like to clarify the new names before making this change. But then happy to have a hand there.

**Comment by ethanfrey** [comment]
Okay, there were a number of naming suggestions in #302, and would like a conclusion that at least @webmaster128 @Isabello and I agree on.

We need new names for:
* `IovWriter` -> this handles all keys and exposes `signAndCommit` as main functionality, we will want to expose this over a socket later (eg. from webapp to chrome extension)
* `UserProfile` -> this provides a uniform interface to the entire collection of keys, Actually as is this is more like a wrapper around a list of KeyringEntries, I would focus on that functionality as a concrete chunk and worry about "address book", etc. in the ui not so in iov-core
* `KeyringEntry` -> this connects to one object that is capable of securely signing bytes and manages multiple private keys.

For me `Wallet` is a complete collection, not a component, and I would prefer to assign that name to either `IovWriter` or `UserProfile`. I also notice other apps (eg gnosis safe) ask you to connect your wallet (where wallet is the remote app that just approves transactions...)

the `LedgerKeyringEntry` is technically a `HardwareSecurityModule`. That is the most standard term for it. Which is why I though `KeyringEntry` -> `SecurityModule` would make sense.

Anyway, above my thoughts, I will make my suggestion below, which can be :+1: or :-1: but I would love to see a counter-proposal at least address these thoughts.

**Comment by ethanfrey** [comment]
Proposal:

`IovWriter` -> `Wallet`
`UserProfile` -> `Keyring`
`***KeyringEntry` -> `***SecurityModule`

**Comment by webmaster128** [comment]
@ethanfrey I think the naming is good, see e.g. from #302

> I agree that "wallet" is a better name for "keyring entry" (which matches https://github.com/lazaridiscom/mm-vault/blob/master/dev/terminology.md)

or https://cdn-images-1.medium.com/max/2000/1*wqZOiTmcCv0ZWocto9esLw.png

While I would not change the name of our keyring for now, I think we got consensus on "wallet". This makes

- A `Keyring` contains multiple `KeyringEntry`s (interface).
- A `Slip10Wallet` is a `KeyringEntry`
- A `Ed25519HdWallet` is a `Slip10Wallet`
- A `Secp256k1HdWallet` is a `Slip10Wallet`
- A `Ed25519SimpleAddressKeyringEntry` is a `KeyringEntry`
- A `HardwareSecurityModule` is a `KeyringEntry`

If we move the term "wallet" 3 levels up, is will never be consistent with the idea of a "HD wallet" from the other resources

**Comment by ethanfrey** [comment]
I like `Keyring` too

What do we call the `IovWriter` interface then?

Ideally after #301 inspired work, we want this to be the main entry point. Eg. a dapp can connect to a `IovWriter` over a socket to `addSignature`, `commitTransaction`, `signAndCommit`, etc. And maybe view `activeIdentity`.

The only vote I have against wallet on this level (and I like those names), is desire to use Wallet on the higher level. What would you propose for `IovWriter`?

**Comment by ethanfrey** [comment]
I would also avoid using both `KeyringEntry` and `Wallet`. People commonly discuss `HardwareWallets`.
In the above case, I would do something like:

* A Keyring contains multiple Wallets (interface).
* A Slip10Wallet is a Wallet
* A Ed25519HdWallet is a Slip10Wallet
* A Secp256k1HdWallet is a Slip10Wallet
* A Ed25519SimpleAddressKeyringEntry is a Wallet (although this class needs cleanup)
* A HardwareWallet is a Wallet

**Comment by Isabello** [comment]
`IovWriter` -> `TransactionAdaptor` 

To me this seems like a logical rename as the Writer looks like its moving to be used only for dealing with transactions, with all of the old connectivity functionality moving into `BcpConnection` and `ChainConnection`.

With this rename, we can use the term `Wallet` for other operations.

**Comment by webmaster128** [comment]
> The only vote I have against wallet on this level (and I like those names), is desire to use Wallet on the higher level. What would you propose for `IovWriter`?

Let me add another perspective: a "wallet" is something, everyone in the industry implements and there is common understanding for it to be a set of addresses and private keys generated from a single seed. What we create around wallets (UserProfile, IovWriter) is something unique with unique capabilities that we should find a proper name for. I am sure, we'll find one.

**Comment by ethanfrey** [comment]
Okay, then we agree on https://github.com/iov-one/iov-core/issues/384#issuecomment-426976595

I am good with that and leave final IovWriter decision for later.
If the relevant PR fulfills part of this, then let's merge it.
